### PR TITLE
Fix control flow analysis in nested loops

### LINKED ISSUE #8418: strictNullChecks with control flow in multiple loops throws TypeError
**TypeScript Version:** 

nightly (1.9.0-dev.20160502)

**Code**

```ts
function test() {
	let x: number | undefined;
	while (y()) {
		while (y()) {
			while (y()) {
				x = foo(x);
			}
		}
		x = 1;
	}
}
function y(): boolean {
	return true;
}
function foo(x: number): number { return 1; }
```

**Expected behavior:**
Compile error reported on `foo(x)`, as `x` can be `undefined` on that location.

**Actual behavior:** 
Type Error:
```
$ ./node_modules/.bin/tsc --strictNullChecks test
node_modules/typescript/lib/tsc.js:35204
                throw e;
                ^

TypeError: Cannot read property 'flags' of undefined
    at isConstEnumObjectType (node_modules/typescript/lib/tsc.js:21310:24)
    at checkExpression (node_modules/typescript/lib/tsc.js:21721:17)
    at checkExpressionWithContextualType (node_modules/typescript/lib/tsc.js:21669:26)
    at checkApplicableSignature (node_modules/typescript/lib/tsc.js:20309:31)
    at chooseOverload (node_modules/typescript/lib/tsc.js:20606:30)
    at resolveCall (node_modules/typescript/lib/tsc.js:20530:26)
    at resolveCallExpression (node_modules/typescript/lib/tsc.js:20669:20)
    at resolveSignature (node_modules/typescript/lib/tsc.js:20791:28)
    at getResolvedSignature (node_modules/typescript/lib/tsc.js:20808:26)
    at checkCallExpression (node_modules/typescript/lib/tsc.js:20824:29)
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #8418. When, during analysis of a back branch, we hit a control flow label that is already in process, we now immediately bail out instead of processing other control flows for that label.

### CODE REVIEW

File: tests/baselines/reference/controlFlowWhileStatement.js [hunk 1]
```diff
@@ -105,6 +105,29 @@ function h3() {
...
     x; // string | number
 }
+// Repro for #8418
+function foo(x: number): number { return 1; }
+function test1() {
+    let x: number | undefined;
+    while (cond) {
+        while (cond) {
+            while (cond) {
+                x = foo(x);
```
Comment (sandersn) [inline tests/baselines/reference/controlFlowWhileStatement.js]: I thought the usage of `foo(x)` was supposed to be a possibly undefined error here.

Comment (mhegazy) [inline tests/baselines/reference/controlFlowWhileStatement.js]: +1

Comment (ahejlsberg) [inline tests/baselines/reference/controlFlowWhileStatement.js]: Ah yes, the error only happens with `--strictNullChecks` and this file doesn't have that option. I will restructure.

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**